Audio

sornas opened the pull request:
## Content

- Over time tweaking of sound effects
- Highpass filter

## TODO

- [x] Helpers for reaching effect X in N seconds (additive)
- [ ] Helpers for reaching effect X in N seconds (multiplicative) (`set_highpass/lowpass_at_time`)
- [x] Documentation

sornas committed:
Add white noise for testing

sornas committed:
Set delay at time

sornas committed:
Refactor

sornas committed:
Add highpass

sornas committed:
Unimplement remove_* on effects

sornas committed:
Documentation

sornas committed:
Change lowpass/highpass over time

sornas committed:
Add tests to main

sornas committed:
Fix tests

sornas committed:
Linear rate of change and exponentials

sornas committed:
const has BIG LETTERS

FredTheDino commented on src/game/game_main.cpp:
```diff
@@ -76,7 +76,7 @@ void setup() {
     };
     Physics::add_shape(LEN(points), points);
-    Renderer::set_window_size(500, 500);
+    Renderer::set_window_size(1000, 1000);
```
You do know that not everyone has a 1000x1000 monitor right?

sornas replied:
Sorry, at 500 I have trouble reading the text. Maybe it could be a command-line flag so we don't edit it back and forth all the time?

FredTheDino replied:
Sounds like we need HighDPI settings

FredTheDino commented on src/engine/util/types.h:
```diff
@@ -35,6 +35,7 @@ struct Image {
 // TODO(ed): Choose a more standard sample rate.
 const u64 AUDIO_SAMPLE_RATE = 48000;
+const u32 AUDIO_SAMPLES_WANT = 2048;
 
```
Maybe these should be somewhere else... I feel like they should live in the mixer header.

sornas replied:
Agreed

FredTheDino commented on src/engine/platform/mixer.h:
```diff
@@ -127,21 +147,33 @@ void stop_sound(AudioID id);
-void Channel::remove_delay();
+// Sets target delay on the channel with the specified settings. The feedback
+// and length is changed over time and reaches their targets after delta_seconds.
+void Channel::set_delay_at_time(f32 feedback, f32 len_seconds, f32 delta_seconds);
```
These could be merged into one function, with default arguments. So something sane for the delta_seconds, because I doubt anyone actually wants it to take effect right at that time, and if they do you can easily specify the delta to be 0.

sornas replied:
Good idea

FredTheDino reviewed (commented):
It looks really good, there are some touchups that can be done here and there. I haven't tested the code though so I trust you it will work, I'll try some things tomarrow but it looks like it should work.

Have you tested the delay lately?

sornas committed:
Code review

- Clamp lowpass/highpass weight.
- Merge set_x and set_x_at_time with default value for the time it
  takes.
- Minor refactor
- Update documentation

sornas commented:
> Have you tested the delay lately?

I hadn't (but I should've), good catch. It worked fine for me.

FredTheDino reviewed (commented):
LGTM
